### PR TITLE
Add ppa:hvr/ghc

### PR DESCRIPTION
**Pull request by sol**
This is required due to
https://github.com/actions/virtual-environments/pull/3268.

### CODE REVIEW

File: setup/src/setup-haskell.ts [hunk 1]
```diff
@@ -24,6 +24,12 @@ export default async function run(
     const os = process.platform as OS;
     const opts = getOpts(getDefaults(os), os, inputs);
 
+    if (os === 'linux') {
+      await exec(
+        'sudo -- sh -c "add-apt-repository -y ppa:hvr/ghc && apt-get update"'
+      );
+    }
```
[the remark this answers is not in the document]
Comment (jared-w) [reply setup/src/setup-haskell.ts:31]: This is an interesting one. Since the PPA is removed by default, it's very costly in terms of time and resources to add it; it also introduces an `apt-get update` which breaks reproducibility for the entire rest of the system. So I'd like to avoid that whenever possible.

And, with that said, hvr's ppa hasn't been updated in 11 weeks and I'm not sure it will continue to be updated, so perhaps it should be opt in rather than implicitly tried after? Having a github action run `apt-get update` for you is quite unusual and shouldn't be something that is reached for so quickly.

There's also the fact that the PPA is now listed on https://www.haskell.org/downloads/ as a secondary option after the now primary option of ghcup, so it seems that ghcup and tools of that nature are going to be more easy to support going forward.

---

tl;dr:

It seems like the best course of action is to actually switch the order of attempts on linux and try using ghcup first, adding the apt repository only if ghcup fails to find a version. That'll keep setup times to a minimum for most cases without breaking existing functionality.

**Comment by ldettwy** [comment]
Just now got this fixed ^^ v1 and v1.2 tags are updated

**Comment by sol** [comment]
@emlisa3162 doesn't seem to work for me.  Here it e.g.

- installs `7.4.1`: https://github.com/sol/haskell-setup-action-bug/runs/4094359642?check_suite_focus=true#step:3:88
- but uses `9.2.1`: https://github.com/sol/haskell-setup-action-bug/runs/4094359642?check_suite_focus=true#step:4:4
